npm/ipsm/ipsm.go

```diff
@@ -124,9 +128,64 @@ func isNsSet(setName string) bool {
 	return !strings.Contains(setName, "-") && !strings.Contains(setName, ":")
 }
 
-// CreateList creates an ipset list. npm maintains one setlist per namespace label.
-func (ipsMgr *IpsetManager) CreateList(listName string) error {
-	if _, exists := ipsMgr.ListMap[listName]; exists {
+// DeleteList removes an ipset list.
+func (ipsMgr *IpsetManager) DeleteList(listName string) error {
+	ipsMgr.Lock()
+	defer ipsMgr.Unlock()
+	return ipsMgr.deleteList(listName)
+}
+
+// DeleteList removes an ipset list.
+func (ipsMgr *IpsetManager) deleteList(listName string) error {
+	entry := &ipsEntry{
+		operationFlag: util.IpsetDestroyFlag,
+		set:           util.GetHashedName(listName),
+	}
+
+	if ipsMgr.listMap[listName].referCount > 0 {
+		ipsMgr.IpSetReferIncOrDec(listName, util.IpsetSetListFlag, DecrementOp)
+		return nil
+	}
+
+	if errCode, err := ipsMgr.run(entry); err != nil {
+		if errCode == 1 {
+			return nil
+		}
+
+		metrics.SendErrorLogAndMetric(util.IpsmID, "Error: failed to delete ipset %s %+v", listName, entry)
+		return err
+	}
+
+	delete(ipsMgr.listMap, listName)
+	return nil
+}
+
+// Run execute an ipset command to update ipset.
+func (ipsMgr *IpsetManager) run(entry *ipsEntry) (int, error) {
+	cmdName := util.Ipset
+	cmdArgs := append([]string{entry.operationFlag, util.IpsetExistFlag, entry.set}, entry.spec...)
+	cmdArgs = util.DropEmptyFields(cmdArgs)
+
+	log.Logf("Executing ipset command %s %v", cmdName, cmdArgs)
+
+	cmd := ipsMgr.exec.Command(cmdName, cmdArgs...)
+	output, err := cmd.CombinedOutput()
+
+	if result, isExitError := err.(utilexec.ExitError); isExitError {
+		exitCode := result.ExitStatus()
+		errfmt := fmt.Errorf("Error: There was an error running command: [%s %v] Stderr: [%v, %s]", cmdName, strings.Join(cmdArgs, " "), err, strings.TrimSuffix(string(output), "\n"))
+		if exitCode > 0 {
+			metrics.SendErrorLogAndMetric(util.IpsmID, errfmt.Error())
+		}
+
+		return exitCode, errfmt
+	}
+
+	return 0, nil
+}
+
+func (ipsMgr *IpsetManager) createList(listName string) error {
+	if _, exists := ipsMgr.listMap[listName]; exists {
```

Comment:
Can we move this check to Public caller ` CreateList(listName string)`, since when we call createList from with the package we are already checking this conditions

Reply:
Can you double check if this change causes an issue in UTs,

Reply:
**createList** is called in multiple places (e.g., . AddToList) So, it is neccessary to keep it in **createList** function. 
I also updated all UTs with this change and all passed, but I will check it again.